### PR TITLE
[BugFix] persist grant/revoke role and support grant impersonate to role

### LINKED ISSUE #10513: Role cannot be granted with impersonate privilege
### Steps to reproduce the behavior (Required)
```sql
mysql> create role testrole;
Query OK, 0 rows affected (0.01 sec)

mysql> grant impersonate on test1 to role testrole;
ERROR 1064 (HY000): Unknown privilege type impersonate
```
### Expected behavior (Required)

Privilege is granted to the role.

### Real behavior (Required)
Unknown privilege error.

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 main

### LINKED ISSUE #10691: The relationship between role and user is invalidated after failover
### Steps to reproduce the behavior (Required)
1. Create a user `test_user` and its role `test_role`.
```
mysql> create role test_role;
Query OK, 0 rows affected (0.03 sec)

mysql> create user test_user;
Query OK, 0 rows affected (0.01 sec)

mysql> grant test_role to test_user;
Query OK, 0 rows affected (0.01 sec)
```

2. Manually failover by restarting the leader of FEs.3. 

3. Grant privileges to the role `test_role`.4. 

```
mysql> grant select_priv on test_priv.test_primary to role 'test_role';
Query OK, 0 rows affected (0.00 sec)

mysql> show grants for test_user;
+---------------------------------+------------------------------------------------------------------------------+
| UserIdentity                    | Grants                                                                       |
+---------------------------------+------------------------------------------------------------------------------+
| 'default_cluster:test_user'@'%' | GRANT Select_priv ON information_schema.* TO 'default_cluster:test_user'@'%' |
+---------------------------------+------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
### Expected behavior (Required)

Privileges is granted to the user `test_user`.

### Real behavior (Required)

Privileges won't be granted to the user `test_user`.

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
main

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [x] bugfix
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10513
Fixes #10691

## Problem Summary(Required) ：

1. Support `GRANT IMPERSONATE ON user TO ROLE role` and `REVOKE IMPERSONATE ON user FROM ROLE role`.
Since the role model will be refactored later, I'm simply restoring impersonate privileges of a role in a simple Set.
2. Persist the relationship between role and user after `GRANT role To user`. The previous implementation will lose this metadata.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 86 / 90 (95.56%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mysql/privilege/RoleManager.java|4|5|80.00%|[130]|
|:large_blue_circle:|com/starrocks/persist/ImpersonatePrivInfo.java|5|6|83.33%|[47]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|39|41|95.12%|[796, 816]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeImpersonateStmt.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|7|7|100.00%|[]|

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 153 / 164 (93.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivTable.java|6|7|85.71%|[77]|
|:large_blue_circle:|com/starrocks/mysql/privilege/RoleManager.java|18|21|85.71%|[135, 249, 250]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|9|10|90.00%|[207]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|74|80|92.50%|[739, 784, 839, 1850, 1851, 1852]|
|:large_blue_circle:|com/starrocks/persist/ImpersonatePrivInfo.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PrivInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeImpersonateStmt.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 159 / 176 (90.34%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|[450, 451, 452, 455, 456, 457]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivTable.java|6|7|85.71%|[77]|
|:large_blue_circle:|com/starrocks/mysql/privilege/RoleManager.java|18|21|85.71%|[135, 249, 250]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|9|10|90.00%|[207]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|76|82|92.68%|[749, 803, 858, 1863, 1864, 1865]|
|:large_blue_circle:|com/starrocks/persist/ImpersonatePrivInfo.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PrivInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeImpersonateStmt.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 159 / 176 (90.34%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|[450, 451, 452, 455, 456, 457]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivTable.java|6|7|85.71%|[77]|
|:large_blue_circle:|com/starrocks/mysql/privilege/RoleManager.java|18|21|85.71%|[135, 249, 250]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|9|10|90.00%|[207]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|76|82|92.68%|[749, 803, 858, 1863, 1864, 1865]|
|:large_blue_circle:|com/starrocks/persist/ImpersonatePrivInfo.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PrivInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeImpersonateStmt.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|

**Comment by padmejin** [comment]
run all

**Comment by padmejin** [comment]
run starrocks_admit_test

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/test/java/com/starrocks/sql/ast/GrantRevokeImpersonateStmtTest.java [hunk 1]
```diff
@@ -70,12 +74,23 @@ public void testNormal() throws Exception {
         com.starrocks.sql.analyzer.Analyzer.analyze(stmt, ctx);
         Assert.assertEquals("GRANT IMPERSONATE ON 'user2'@'%' TO 'user1'@'%'", stmt.toString());
 
+        stmt = (GrantImpersonateStmt) com.starrocks.sql.parser.SqlParser.parse(
+                "grant IMPERSONATE on user2 to ROLE role1", 1).get(0);
+        com.starrocks.sql.analyzer.Analyzer.analyze(stmt, ctx);
+        Assert.assertEquals("GRANT IMPERSONATE ON 'default_cluster:user2'@'%' TO ROLE 'default_cluster:role1'", stmt.toString());
```
Comment (gengjun-git) [inline fe/fe-core/src/test/java/com/starrocks/sql/ast/GrantRevokeImpersonateStmtTest.java:80]: You should rebase the latest main and remove the default_cluster prefix. Or this ut will fail when merge into main branch.

Comment (padmejin) [reply fe/fe-core/src/test/java/com/starrocks/sql/ast/GrantRevokeImpersonateStmtTest.java]: rebased & fixed.

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 159 / 176 (90.34%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|[450, 451, 452, 455, 456, 457]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivTable.java|6|7|85.71%|[77]|
|:large_blue_circle:|com/starrocks/mysql/privilege/RoleManager.java|18|21|85.71%|[135, 249, 250]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|9|10|90.00%|[208]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|76|82|92.68%|[749, 803, 858, 1863, 1864, 1865]|
|:large_blue_circle:|com/starrocks/persist/ImpersonatePrivInfo.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PrivInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeImpersonateStmt.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by chaoyli — APPROVED** [review]

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
@Mergifyio backport branch-2.4 branch-2.3
¡Muchas gracias!

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 159 / 176 (90.34%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|[450, 451, 452, 455, 456, 457]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivTable.java|6|7|85.71%|[77]|
|:large_blue_circle:|com/starrocks/mysql/privilege/RoleManager.java|18|21|85.71%|[135, 249, 250]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|9|10|90.00%|[208]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|76|82|92.68%|[749, 803, 858, 1863, 1864, 1865]|
|:large_blue_circle:|com/starrocks/persist/ImpersonatePrivInfo.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/PrivInfo.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeImpersonateStmt.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeImpersonateStmt.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|

**Pull request merged**